clients/src/main/java/org/apache/kafka/common/network/PlaintextTransportLayer.java

```diff
@@ -100,7 +100,13 @@ public void handshake() {}
     */
     @Override
     public int read(ByteBuffer dst) throws IOException {
-        return socketChannel.read(dst);
+        int readSize = socketChannel.read(dst);
+
+        if (!completedRead && ((ByteBuffer) dst.rewind()).getInt() > 1_000_000)  // first responses should negotiate API_VERSION and be small in size
+            throw new IllegalTransportLayerStateException("Received a first response larger than 1MB (Is this a plaintext response?)");
```

Comment:
For your consideration: message could be more direct .

`Is this a plaintext response?` --> `The broker expects SSL, is your client configured for SSL correctly?` (or something of that nature)

Reply:
At this layer we don't know whether the broker expects SSL. The broker is actually configured to expect plaintext in this scenario and we don't know whether the response here is SSL. As @ijuma said earlier, that is a layering violation so I think it's better to keep the current message

Reply:
Ack. Any other suggestions? Would something like `(Please verify that the configuration is correct. It may be that we are not connecting to a plaintext port)` work better?

Reply:
Adding an extra "if" statement on every call to `read()` seems like a very heavy price to pay for this feature.  We should look at where we're using the total length and check there, I think.

Reply:
I like that message better. I'm trying to solve a similar problem- connecting to the C3 back end instead of Kafka - C3 responds with HTTP which is like 1.3 GB allocation on the heap. So a prompt to check your config and check you're connecting to the correct host/port would be good.
https://confluent.slack.com/archives/CG3B0PUM7/p1572450226012300?thread_ts=1572449855.012200&cid=CG3B0PUM7

Reply:
@cmccabe Doesn't the compiler optimize the if check? 

I do not know how to proceed with this change.

@astubbs's example wouldn't work for the common case where we try to allocate ~106MBs and it would also be appleid to every receive we get, whereas here we are trying to scope it to the first receive for a specific channel

Reply:
Are you referring to my fix? or my problem? I just would like my problem solved :) I'm not concerned about my fix - I was just mainly trying to demonstrate the problem.

Reply:
If you wanted to minimize the cost, you could move the exception and second part of the if to a separate method so that it's not in the hot path. Btw, calling `rewind` on the `ByteBuffer` mutates it, which seems generally unsafe?

Reply:
Also, I thought the previous discussion indicates that 1 MB might be too low of a check?

Reply:
I think the right time to check this would be when we originally create the NetworkReceive object, not in the hot path for every read.  Also, whatever we choose as the "upper limit" for a size to be interpreted as valid needs to become the new upper limit for `message.max.bytes` (which in turn, means that maybe this needs a KIP, since it changes a public config)